js/ui/payg.js

```diff
@@ -212,3 +219,249 @@ class PaygUnlockUi extends St.Widget {
+    }
+
+    entryReset() {
+        this._codeEntry.set_text('');
+    }
+
+    entrySetEnabled(enabled) {
+        if (this._codeEntry._enabled === enabled)
+            return;
+
+        this._codeEntry._enabled = enabled;
+        this._codeEntry.reactive = enabled;
+        this._codeEntry.can_focus = enabled;
+        this._codeEntry.clutter_text.reactive = enabled;
+        this._codeEntry.clutter_text.editable = enabled;
+        this._codeEntry.clutter_text.cursor_visible = enabled;
+    }
+
+    get entryCode() {
+        return this._codeEntry.get_text();
+    }
+
+    get verificationStatus() {
+        return this._verificationStatus;
+    }
+
+    set verificationStatus(value) {
+        this._verificationStatus = value;
+    }
+
+    get spinner() {
+        return this._spinner;
+    }
+
+    get applyButton() {
+        return this._applyButton;
+    }
+
+    get buttonBox() {
+        return this._buttonBox;
+    }
+
+});
+
+var ApplyCodeNotification = GObject.registerClass(
+class ApplyCodeNotification extends MessageTray.Notification {
+    _init(source, title, banner) {
+        super._init(source, title, banner);
+
+        this._titleOrig = title;
+
+        // Note: "banner" is actually the string displayed in the banner, not a
+        // banner object. This variable name simply follows the convention of
+        // the parent class.
+        this._bannerOrig = banner;
+        this._verificationStatus = UnlockStatus.NOT_VERIFYING;
+    }
+
+    createBanner() {
+        this._banner = new MessageTray.NotificationBanner(this);
+        this._unlockWidget = new PaygUnlockWidget();
+        this._unlockWidget.connect('code-added', this._onCodeAdded.bind(this));
+        this._unlockWidget.connect('code-rejected', this._onCodeRejected.bind(this));
+        this._banner.setActionArea(this._unlockWidget.buttonBox);
+
+        return this._banner;
+    }
+
+    _onCodeAdded() {
+        this._setMessage(_('Code applied successfully!'));
+
+        GLib.timeout_add_seconds(
+            GLib.PRIORITY_DEFAULT,
+            SUCCESS_DELAY_SECONDS,
+            () => {
+                this.emit('done-displaying');
+                this.destroy();
+
+                return GLib.SOURCE_REMOVE;
+            });
+    }
+
+    // if errorMessage is unspecified, a default message will be populated based
+    // on whether time remains
+    _onCodeRejected(unlockWidget, errorMessage) {
+        this._setMessage(errorMessage ? errorMessage : this._bannerOrig);
+    }
+
+    _setMessage(message) {
+        this.update(this._titleOrig, message);
+    }
+
+    activate() {
+        // We get here if the Apply button is inactive when we try to click it.
+        // Unless we're already done, exit early so we don't destroy the
+        // notification)
+        if (this._verificationStatus !== UnlockStatus.SUCCEEDED)
+            return;
+
+        super.activate();
+    }
+});
+
+// Takes an UNIX timestamp (in seconds) and returns a string
```

Comment:
It's really just a number of seconds not a UNIX timestamp, perhaps we should update the comment.